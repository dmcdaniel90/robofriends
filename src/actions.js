import { CHANGE_SEARCH_FIELD } from './constants.js';
import { REQUEST_ROBOTS_PENDING } from './constants.js';
import { REQUEST_ROBOTS_SUCCESS } from './constants.js';
import { REQUEST_ROBOTS_FAILED } from './constants.js';

export const setSearchField = (text) => ({
  type: CHANGE_SEARCH_FIELD,
  payload: text
})

export const requestRobots = () => (dispatch) => {
  dispatch({ type: REQUEST_ROBOTS_PENDING });
  fetch('https://jsonplaceholder.typicode.com/users', {
    headers : { 
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      'User-Agent': 'PokeFriends (localhost:3000/robofriends, v1.0.0)'
     }
  })
  
    .then(response => response.json())
    .then(data => dispatch({ type: REQUEST_ROBOTS_SUCCESS, payload: data }))
    .catch(error => dispatch({ type: REQUEST_ROBOTS_FAILED, payload: error }))
}

