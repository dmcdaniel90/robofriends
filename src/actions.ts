import { CHANGE_SEARCH_FIELD } from './constants.ts';
import { REQUEST_ROBOTS_PENDING } from './constants.ts';
import { REQUEST_ROBOTS_SUCCESS } from './constants.ts';
import { REQUEST_ROBOTS_FAILED } from './constants.ts';
import { apiCall } from './api/api.ts';
import { Robot, Action } from './types/common.ts';

export const setSearchField = (text: string) => <Action> ({
  type: CHANGE_SEARCH_FIELD,
  payload: text
})

export const requestRobots = () => async (dispatch: (arg: Action) => void) => {
  dispatch({ type: REQUEST_ROBOTS_PENDING });
  try {
    const data = await apiCall<Robot>('https://jsonplaceholder.typicode.com/users');
    dispatch({ type: REQUEST_ROBOTS_SUCCESS, payload: data });
  } catch (error: unknown) {
    dispatch({ type: REQUEST_ROBOTS_FAILED, payload: error });
  }
}

