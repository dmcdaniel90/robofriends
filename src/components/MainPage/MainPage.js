import React, { useEffect } from "react";
import Header from "../Header/Header";
import CardList from "../CardList/CardList";
import SearchBox from "../SearchBox/SearchBox";
import Scroll from "../Scroll/Scroll";
import ErrorBoundary from "../ErrorBoundary/ErrorBoundary";
import './MainPage.css';


function MainPage(props) {
  const { searchField, onSearchChange, robots, onRequestRobots, isPending, error } = props;

  useEffect(() => {
    onRequestRobots();
  }, [onRequestRobots]);

  const filterRobots = () => {
    return robots.filter(robot => {
      return robot.name.toLowerCase().includes(searchField.toLowerCase())
    })
  }

  return isPending ?
    <h1 className = "f1 abs-centered" id="loading-screen">Loading</h1> :
      (
        <div className="tc" data-testid="mainpage">
          <Header />
          <SearchBox searchChange={onSearchChange} />
          <Scroll>
            <ErrorBoundary error={error}>
              <CardList robots={filterRobots()} />
            </ErrorBoundary>
          </Scroll>
        </div>
      ) 
  }

export default MainPage;
