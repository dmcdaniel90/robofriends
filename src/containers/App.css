@font-face {
    font-family: 'SEGA LOGO FONT';
    font-style: normal;
    font-weight: normal;
    src: local('SEGA LOGO FONT'), url('SEGA.woff') format('woff');
    }

h1 {
    font-family: 'SEGA LOGO FONT';
    font-weight: 200;
    color: #0ccac4;
}

.abs-centered {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  z-index: -1;
}