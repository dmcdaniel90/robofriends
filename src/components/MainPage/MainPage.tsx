import { useEffect } from "react";
import Header from "../Header/Header";
import CardList from "../CardList/CardList";
import SearchBox from "../SearchBox/SearchBox";
import Scroll from "../Scroll/Scroll";
import ErrorBoundary from "../ErrorBoundary/ErrorBoundary";
import { Robot } from "../../types/common"; 
import './MainPage.css';

type MainPageProps = {
  searchField: string,
  onSearchChange: (event: React.SyntheticEvent<HTMLInputElement>) => void,
  robots: Array<Robot>,
  onRequestRobots: () => void,
  isPending: boolean,
  error: Error
}


function MainPage(props: MainPageProps) {
  const { searchField, onSearchChange, robots, onRequestRobots, isPending, error } = props;

  useEffect(() => {
    onRequestRobots();
  }, [onRequestRobots]);

  const filterRobots = (): Robot[] => {
    return robots.filter(robot => {
      return robot.name.toLowerCase().includes(searchField.toLowerCase())
    })
  }

  return isPending ?
    <h1 className = "f1 abs-centered" id="loading-screen">Loading</h1> :
      (
        <div className="tc" data-testid="mainpage">
          <Header />
          <SearchBox searchChange={onSearchChange} />
          <Scroll>
            <ErrorBoundary error={error}>
              <CardList robots={filterRobots()} />
            </ErrorBoundary>
          </Scroll>
        </div>
      ) 
  }

export default MainPage;
