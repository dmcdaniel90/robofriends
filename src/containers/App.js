import { connect } from 'react-redux';
import { setSearchField, requestRobots } from "../actions";
import MainPage from "../components/MainPage/MainPage";

const mapStateToProps = (state) => {
  return {
    searchField: state.searchRobots.searchField,
    robots: state.requestRobots.robots,
    isPending: state.requestRobots.isPending,
    error: state.requestRobots.error
  }
}

export const mapDispatchToProps = (dispatch) => { //exported for testing
  return {
    onSearchChange: (event) => dispatch(setSearchField(event.target.value)),
    onRequestRobots: () => dispatch(requestRobots())
  }
}

function App(props) {
  return <MainPage {...props} />
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
