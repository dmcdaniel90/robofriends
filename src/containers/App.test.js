import { mapDispatchToProps } from './App';
import { render } from '@testing-library/react';
import * as actions from '../actions';
import MainPage from '../components/MainPage/MainPage';

it('dispatches setSearchField action on search box change', () => {
  const mockDispatch = jest.fn();
  const mockSetSearchField = jest.spyOn(actions, 'setSearchField');
  const mockEvent = { target: { value: 'test' } };

  mapDispatchToProps(mockDispatch).onSearchChange(mockEvent);

  expect(mockSetSearchField).toHaveBeenCalledWith('test');
  expect(mockDispatch).toHaveBeenCalledTimes(1);
})

it('dispatches requestRobots action on component mount', () => {
  const mockDispatch = jest.fn();
  const mockRequestRobots = jest.spyOn(actions, 'requestRobots');

  mapDispatchToProps(mockDispatch).onRequestRobots();

  expect(mockRequestRobots).toHaveBeenCalled();
  expect(mockDispatch).toHaveBeenCalledTimes(1);
})

it('renders the MainPage component', () => {
  const mockProps = {
    onRequestRobots: jest.fn(),
    onSearchChange: jest.fn(),
    robots: [],
    searchField: '',
    isPending: false
  }

  const {container} = render(<MainPage {...mockProps} />)
  expect(container).toMatchSnapshot();
})