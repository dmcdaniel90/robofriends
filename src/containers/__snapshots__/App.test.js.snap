// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the MainPage component 1`] = `
<div>
  <div
    class="tc"
  >
    <h1
      class="f1"
    >
      RoboFriends
    </h1>
    <div
      class="pa2"
      data-testid="search-box"
    >
      <label
        for="search"
        hidden=""
      >
        Search Robots
      </label>
      <input
        class="pa3 ba b--green bg-lightest-blue"
        id="search"
        placeholder="Search Robots"
        type="search"
      />
    </div>
    <div
      style="overflow-y: scroll; border: 3px solid black; height: 800px;"
    >
      <div
        data-testid="card-list"
      />
    </div>
  </div>
</div>
`;
